File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/validation/NotInValidatorTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.Collections;
import java.util.Set;

import org.junit.jupiter.api.Test;

class NotInValidatorTest {

    @Test
    void testNullValue() {
        NotInValidator validator = new NotInValidator(Collections.singleton("invalid"));
        assertFalse(validator.eval(null), "Null value should not be allowed");
    }

    @Test
    void testEmptyInvalidValues() {
        NotInValidator validator = new NotInValidator(Collections.emptySet());
        assertTrue(validator.eval("value"), "Empty sets should allow any value");
    }

    @Test
    void testValueNotInInvalidList() {
        NotInValidator validator = new NotInValidator(Collections.singleton("invalid"));
        assertTrue(validator.eval("valid"), "Values not in the invalid list should be allowed");
    }

    @Test
    void testValueInInvalidList() {
        NotInValidator validator = new NotInValidator(Collections.singleton("invalid"));
        assertFalse(validator.eval("invalid"), "Values in the invalid list should not be allowed");
    }

    @Test
    void testMultipleInvalidValues() {
        NotInValidator validator = new NotInValidator(Set.of("invalid1", "invalid2", "invalid3"));
        assertTrue(validator.eval("valid"));
        assertFalse(validator.eval("invalid1"));
        assertFalse(validator.eval("invalid2"));
        assertFalse(validator.eval("invalid3"));
    }

    @Test
    void testSerialization() throws Exception {
        NotInValidator validator = new NotInValidator(Set.of("invalid1", "invalid2"));
        ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
        ObjectOutputStream out = new ObjectOutputStream(byteOut);
        out.writeObject(validator);

        ByteArrayInputStream byteIn = new ByteArrayInputStream(byteOut.toByteArray());
        ObjectInputStream in = new ObjectInputStream(byteIn);
        NotInValidator deserializedValidator = (NotInValidator) in.readObject();

        assertEquals(validator.eval("valid"), deserializedValidator.eval("valid"));
        assertFalse(deserializedValidator.eval("invalid1"));
        assertFalse(deserializedValidator.eval("invalid2"));
    }
}

File: lib/python/opentoken/src/test/opentoken/attributes/validation/not_in_validator_test.py
"""
Tests for NotInValidator.
"""

import pickle
import pytest

from opentoken.attributes.validation.not_in_validator import NotInValidator


class TestNotInValidator:
    """Test cases for NotInValidator."""

    def test_null_value(self):
        """Test that null (None) values are not allowed."""
        validator = NotInValidator({"invalid"})
        assert validator.eval(None) is False, "Null value should not be allowed"

    def test_empty_invalid_values(self):
        """Test that an empty set allows any value."""
        validator = NotInValidator(set())
        assert validator.eval("value") is True, "Empty sets should allow any value"

    def test_value_not_in_invalid_list(self):
        """Test that values not in the invalid list are allowed."""
        validator = NotInValidator({"invalid"})
        assert validator.eval("valid") is True, "Values not in the invalid list should be allowed"

    def test_value_in_invalid_list(self):
        """Test that values in the invalid list are not allowed."""
        validator = NotInValidator({"invalid"})
        assert validator.eval("invalid") is False, "Values in the invalid list should not be allowed"

    def test_multiple_invalid_values(self):
        """Test that multiple invalid values are handled correctly."""
        validator = NotInValidator({"invalid1", "invalid2", "invalid3"})
        assert validator.eval("valid") is True, "Values not in the invalid list should be allowed"
        assert validator.eval("invalid1") is False, "Values in the invalid list should not be allowed"
        assert validator.eval("invalid2") is False, "Values in the invalid list should not be allowed"
        assert validator.eval("invalid3") is False, "Values in the invalid list should not be allowed"

    def test_serialization(self):
        """Test serialization and deserialization of the validator."""
        validator = NotInValidator({"invalid1", "invalid2"})

        # Serialize the validator using pickle
        serialized_data = pickle.dumps(validator)

        # Deserialize the validator
        deserialized_validator = pickle.loads(serialized_data)

        # Test that both validators behave identically
        assert validator.eval("valid") == deserialized_validator.eval("valid"), (
            "Validation results should match for 'valid'"
        )
        assert deserialized_validator.eval("invalid1") is False, (
            "Deserialized validator should not allow 'invalid1'"
        )
        assert deserialized_validator.eval("invalid2") is False, (
            "Deserialized validator should not allow 'invalid2'"
        )